Add tests for extracted lib and loose application EAR

tjwatson opened the pull request:
- [ ] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [ ] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [ ] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

tjwatson committed:
Add tests for extracted lib and loose application EAR

tjwatson commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=37b2a5e5-d8ed-4d34-a98f-3479aa3c0c2b) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

jhanders34 commented on dev/com.ibm.ws.classloading_sharedclass_fat/publish/servers/sharedClassesEarLooseTest/apps/sharedClassesEarTest.ear.xml:
```diff
@@ -0,0 +1,40 @@
+			<dir sourceOnDisk="${server.config.dir}/looseContent/sharedClassesWarLib"
+				targetInArchive="/" />
+		</archive>
+		<dir sourceOnDisk="${server.config.dir}/looseContent/sharedClassesWarTest/WEB-INF/classes"
+			targetInArchive="/WEB-INF/classes" />
```
```suggestion
		<dir sourceOnDisk="${server.config.dir}/looseContent/sharedClassesWarTest/WEB-INF/classes"
			targetInArchive="/WEB-INF/classes" />
```

The WEB-INF/classes does not need to be part of the sourceOnDisk.  I can be any directory and then you map it to WEB-INF/classes.  It  would be good to have that scenario.

It would also be good to have the scenario where you have looseContent/sharedClassesWarTest/a/A.class and possibly similarly b/B.class and then you have the targetInArchive to be WEB-INF/classes/restOfPackage".  In this case we want the URL for shared class cache to be the entry URL and not the container URL since the location of the class doesn't match the package, so it needs to be the entry.  This is the main scenario I was concerned with in your updates to the shared class cache URL.

jhanders34 replied:
I believe having multiple dirs in the archive would be good since it will drive additional URLs in the code to cause the container to not be only one URL like is in the other scenarios if it is a zip for instance.

tjwatson replied:
> It would also be good to have the scenario where you have looseContent/sharedClassesWarTest/a/A.class and possibly similarly b/B.class and then you have the targetInArchive to be WEB-INF/classes/restOfPackage"

This fails with the current code, the class is never stored in the class cache because null is returned from here: https://github.com/OpenLiberty/open-liberty/blob/8947456f91dd71b8e77717cef80245a05ede1ee3/dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ContainerClassLoader.java#L289

The result is we never store classes with this scenario.  Is that your expectation?

> I believe having multiple dirs in the archive would be good since it will drive additional URLs in the code to cause the container to no be only one URL like is in the other scenarios if it is a zip for instance.

I'll have to try this out but the directory structure for the package name will have to be kept intact to make the current code store the class.  So it will be something like this:

```xml
		<dir sourceOnDisk="${server.config.dir}/looseContent/warPkgA"
			targetInArchive="/WEB-INF/classes" />
		<dir sourceOnDisk="${server.config.dir}/looseContent/warPkgB"
			targetInArchive="/WEB-INF/classes" />
```
Where `warPkgA` and `warPkgB` folders still contain the package name folder hierarchy.

jhanders34 replied:
I wonder if I tested that out and it didn't work when making an update or something.  It would be interesting to change that to use the entry URL to see if it works.  Interestingly we do not do a null URL for when it is in a jar file in a different directory.

tjwatson replied:
Is that even a scenario that is worth storing the classes for?  It seems very advanced and I have doubts the liberty tooling produce loose XML files that do that.  I say that because I think most any Java compiler is going to place `.class` files into an output folder hierarchy with the package name intact. Then the tooling would produce a loose XML file pointing to the base output folder on disk.

tjwatson replied:
>I believe having multiple dirs in the archive would be good since it will drive additional URLs in the code to cause the container to not be only one URL like is in the other scenarios if it is a zip for instance.

I added this to the test.  This scenario did fail for my pr #30945.  I've now updated that PR to pass the test.

Still need to decide what the behavior should be for the case where loose content on disk does not have matching package name hierarchy folders.

jhanders34 replied:
Yeah I think it is ok not to put them in the shared class cache like what was done before, but we still need a test to show that it does not get put into the shared class cache and we don't try to find it in the shared class cache because the URL is null.

tjwatson replied:
Added new test for that

tjwatson committed:
Use multiple sources for WEB-INF/classes

tjwatson committed:
Add trace and test for when there is no valid shared class URL

If the class content is in a folder the class must be contained
in a folder hierarchy with the package name. Added trace messages
for when this is not the case and added a test to exercise the
new trace.

tjwatson commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=ec709d7c-8306-4ce9-a1f1-b44fbc8e3b35) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tjwatson committed:
Add another test that has one valid path for sharing and one invalid

tjwatson commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=82dea09c-5750-49e7-b136-d604fab910e5) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

jhanders34 commented on dev/com.ibm.ws.classloading_sharedclass_fat/publish/servers/sharedClassesEarLooseTest/apps/sharedClassesEarTest.ear.xml:
```diff
@@ -0,0 +1,52 @@
+    SPDX-License-Identifier: EPL-2.0
+   
+    Contributors:
+        IBM Corporation - initial API and implementation
```
```suggestion
```

Not needed in new files.

tjwatson committed:
Remove contributors section from copyright

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

tjwatson commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=bad23031-384c-4a83-ae89-7fca93500110) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._
